pkg/reconciler/pipelinerun/resources/apply_test.go

```diff
@@ -2034,6 +2034,285 @@ func TestApplyParameters_ArrayIndexing(t *testing.T) {
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("$(params.first-param)"),
+						}, {
+							Name: "second-param", Value: *v1beta1.NewStructuredValues("$(params.second-param)"),
+						}, {
+							Name: "third-param", Value: *v1beta1.NewStructuredValues("new-value"),
+						}}},
+					}},
+			}},
+		},
+		params: v1beta1.Params{{Name: "second-param", Value: *v1beta1.NewStructuredValues("second-value")}},
+		expected: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeString, Default: v1beta1.NewStructuredValues("default-value"),
+			}, {
+				Name: "second-param", Type: v1beta1.ParamTypeString,
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("default-value"),
+						}, {
+							Name: "second-param", Value: *v1beta1.NewStructuredValues("second-value"),
+						}, {
+							Name: "third-param", Value: *v1beta1.NewStructuredValues("new-value"),
+						}}},
+					}},
+			}},
+		},
+	}, {
+		name: "matrix include params - object replacement",
+		original: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "myobject",
+				Properties: map[string]v1beta1.PropertySpec{
+					"key1": {Type: "string"},
+					"key2": {Type: "string"},
+				},
+				Default: v1beta1.NewObject(map[string]string{
+					"key1": "default",
+					"key2": "param!",
+				}),
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("$(params.myobject.key1)"),
+						}, {
+							Name: "second-param", Value: *v1beta1.NewStructuredValues("$(params.myobject.key2)"),
+						}}},
+					}},
+			}},
+		},
+		params: v1beta1.Params{{Name: "myobject", Value: *v1beta1.NewObject(map[string]string{
+			"key1": "pipeline",
+			"key2": "param!!",
+		})}},
+		expected: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "myobject",
+				Properties: map[string]v1beta1.PropertySpec{
+					"key1": {Type: "string"},
+					"key2": {Type: "string"},
+				},
+				Default: v1beta1.NewObject(map[string]string{
+					"key1": "default",
+					"key2": "param!",
+				}),
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("pipeline"),
+						}, {
+							Name: "second-param", Value: *v1beta1.NewStructuredValues("param!!"),
+						}},
+					},
+					}},
+			}},
+		},
+	}, {
```

Comment:
is this example demonstrating object replacements or string replacements?

looking at the code where we fetch the replacement, specific values in the object are string replacements: https://github.com/tektoncd/pipeline/blob/b7e49f2d2f85e74fc88f3099b73b87fadd482616/pkg/reconciler/pipelinerun/resources/apply.go#L103-L137

Reply:
let's keep this test case, but clarify that it's testing string replacements from object params

Reply:
Remove object test altogether

Reply:
please bring it back, it was a useful test (and confirms our understanding of replacements)

Reply:
Sorry, didn't see the first comment until after. Restored and renamed the test to "matrix include params - string replacement from object params""